File: ai.py
import random, copy
from tqdm import tqdm

from board import Board, Field, Pawn
from moves import Move, PlacePawn, MovePawnAndWall, PlaceWall, MovePawn, GameEnd
from rules import Referee, findValidPath


AREA_COEF = 1
FREEMOV_COEF = 0.3
PAWN_BARRIER_COEF = 0.5


class TreeNode():
    
    def __init__(self, children: list['TreeNode'], grade: float, board: Board, is_base: bool = False) -> None:
        self.children: list['TreeNode'] = children
        self.grade: float = grade
        self.board: Board = board
        self.is_base: bool = is_base

    def setChildren(self, children: list['TreeNode']) -> None:
        self.children = children
    
    def setGrade(self, grade: float) -> None:
        self.grade = grade
        
    def setBoard(self, board: Board) -> None:
        self.board = board
    
    def isBase(self) -> bool:
        return self.is_base

    def isLeaf(self) -> bool:
        return len(self.children) == 0
    
    def print(self) -> None:
        self.visualize()
        
    def visualize(self, level=0):
        indent = " " * (level * 4)
        print(f"{indent}Grade: {self.grade}, Is Base: {self.is_base}")
        for child in self.children:
            if isinstance(child, list):
                for sub_child in child:
                    sub_child.visualize(level + 1)
            else:
                child.visualize(level + 1)

class Fendoter():
    
    def __init__(self, player, playing_method: str, search_depth) -> None:
        self.ref = Referee()
        self.player = player
        self.opponent = 3 - player #TODO: hardcoded, only works for 2 player
        self.playing_method = playing_method
        self.search_depth = search_depth
        self.search_tree: TreeNode = TreeNode([], 0, None, True)
    
    def makeMove(self, board: Board) -> Move:
        best_move: Move = self.evaluateMoves(board, self.playing_method)
        #self.search_tree.print() # debug
        return best_move
    
    def evaluateMoves(self, board: Board, method: str) -> Move:
        match method:
            case "depth1":
                return self.depth1Eval(board)
            case "random":
                return self.playRandom(board)
            case "minimax":
                move, grade, self.search_tree = self.minimax(board, depth=self.search_depth, maximizing_player=True)
                return move
            case "negamax":
                move, grade, self.search_tree = self.negamax(board, depth=self.search_depth, p=1)
                return move
            case "alpha-beta":
                move, grade , self.search_tree = self.alphabeta(board, depth=self.search_depth, alpha=float('-inf'), beta=float('inf'), p=1)
                return move
            case _:
                raise ValueError("Invalid playing method")
    
    
    def calculateMoves(self, board: Board) -> tuple[list[Move], list[Board]]:
        # debug start
        #print("\nCurrent Board:\n")
        #print(board)
        # debug end
        
        own_pawns: list[Pawn] = board.getPawns(board.getTurn())
        fields: list[Field] = board.getFieldsFlat()
        board_state: dict = board.getState()

        moves = []
        new_boards = []
        for pawn in own_pawns:
            for field in fields:
                # Place new Pawn
                if self.ref.checkLegalMove(PlacePawn(field.getCoordinates(), board.getTurn()), board_state):
                    moves.append(PlacePawn(field.getCoordinates(), board.getTurn()))
                    new_board = Board(board_state['size'], board_state['max_pawns'], new=False)
                    #new_board.loadState(copy.deepcopy(board_state))
                    new_board.loadState(board_state)
                    new_board.placePawn(field.getCoordinates(), board.getTurn())
                    new_board.endTurn()
                    new_boards.append(new_board)
                for direction in ["N", "E", "S", "W"]:
                    # Move Pawn and place Wall
                    if self.ref.checkLegalMove(MovePawnAndWall(pawn.getCoordinates(), field.getCoordinates(), direction, board.getTurn()), board_state):
                        moves.append(MovePawnAndWall(pawn.getCoordinates(), field.getCoordinates(), direction, board.getTurn()))
                        new_board = Board(board_state['size'], board_state['max_pawns'], new=False)
                        #new_board.loadState(copy.deepcopy(board_state))
                        new_board.loadState(board_state)
                        new_board.movePawn(pawn.getCoordinates(), field.getCoordinates())
                        new_board.placeWall(field.getCoordinates(), direction, board.getTurn())
                        new_board.endTurn()
                        new_boards.append(new_board)
                    # Place Wall without moving Pawn
                    if self.ref.checkLegalMove(PlaceWall(field.getCoordinates(), direction, board.getTurn()), board_state):
                        moves.append(PlaceWall(field.getCoordinates(), direction, board.getTurn()))
                        new_board = Board(board_state['size'], board_state['max_pawns'], new=False)
                        #new_board.loadState(copy.deepcopy(board_state))
                        new_board.loadState(board_state)
                        new_board.placeWall(field.getCoordinates(), direction, board.getTurn())
                        new_board.endTurn()
                        new_boards.append(new_board)
                        
        return moves, new_boards
    
    def depth1Eval(self, board: Board) -> Move:
        moves, boards = self.calculateMoves(board)
        best_move = boards[0]
        best_grade = self.grade(best_move)
        for state in boards:
            grade = self.grade(state)
            if grade > best_grade:
                best_move = state
                best_grade = grade
        best_move_index = boards.index(best_move)
        return moves[best_move_index]
        
    
    def playRandom(self, board: Board) -> int:
        possible_moves, new_boards = self.calculateMoves(board)
        return random.choice(possible_moves)

    
    def grade(self, board: Board) -> int:
        ''' Takes a board and returns the grade for the current player.'''

        board.evaluateFields() # TODO: improve evaluation performance
        if board.getWinner() == self.player:
            return float('inf')
        if board.getWinner() == self.opponent:
            return float('-inf')

        current_pawns = board.getPawns(self.player)
        opponent_pawns = board.getPawns(self.opponent)

        # grade area
        area_grade = board.getPlayerArea(self.player) - board.getPlayerArea(self.opponent)

        # grade movement freedom
        # freedom_grade = self.calculateMoves(board)[0] # prossessing time too long
        freedom_grade, current_player_freedom_grade, opponent_freedom_grade = 0, 0, 0
        for direction in ["N", "E", "S", "W"]: # estimate freedom by checking walls/pawns/boarders next to pawns
            for pawn in current_pawns:
                if (pawn.getCoordinates()[0] == 0 and direction == "W") or (pawn.getCoordinates()[0] == board.getSize() - 1 and direction == "E") or (pawn.getCoordinates()[1] == 0 and direction == "N") or (pawn.getCoordinates()[1] == board.getSize() - 1 and direction == "S"):
                    current_player_freedom_grade -=1
                elif board.getFields()[pawn.getCoordinates()].getWall(direction):
                    current_player_freedom_grade -= 1
                else:
                    end_coords = board.getFields()[pawn.getCoordinates()].getNeighborCoords(direction)
                    if end_coords:
                        if board.getField(end_coords).getPawn():
                            current_player_freedom_grade -= PAWN_BARRIER_COEF*1
                # different approach (but more computing heavy):
                #end_coords = board.getFields()[pawn.getCoordinates()].getNeighborCoords(direction)
                #if end_coords:
                #    if findValidPath(pawn.getCoordinates(), end_coords, board.getFields()):
                #        current_player_freedom_grade += 1
            for pawn in opponent_pawns:
                if (pawn.getCoordinates()[0] == 0 and direction == "W") or (pawn.getCoordinates()[0] == board.getSize() - 1 and direction == "E") or (pawn.getCoordinates()[1] == 0 and direction == "N") or (pawn.getCoordinates()[1] == board.getSize() - 1 and direction == "S"):
                    opponent_freedom_grade -=1
                elif board.getFields()[pawn.getCoordinates()].getWall(direction):
                    opponent_freedom_grade -= 1
                else:
                    end_coords = board.getFields()[pawn.getCoordinates()].getNeighborCoords(direction)
                    if end_coords:
                        if board.getField(end_coords).getPawn():
                            opponent_freedom_grade -= PAWN_BARRIER_COEF*1
                # different approach (but more computing heavy):
                # end_coords = board.getFields()[pawn.getCoordinates()].getNeighborCoords(direction)
                # if end_coords:
                #     if findValidPath(pawn.getCoordinates(), end_coords, board.getFields()):
                #         opponent_freedom_grade += 1

        freedom_grade = (current_player_freedom_grade / len(current_pawns)) - (opponent_freedom_grade / len(opponent_pawns))
        
        grade = AREA_COEF * area_grade + FREEMOV_COEF * freedom_grade
        return grade
        
    def minimax(self, board: Board, depth: int, maximizing_player: bool) -> tuple[int, list[TreeNode]]:
        if depth == 0:
            grade = self.grade(board)
            return None, grade, TreeNode([], grade, board)
        
        children = []
        if maximizing_player:
            max_eval = float('-inf')
            next_moves, new_boards = self.calculateMoves(board)
            for new_board in new_boards:
                _, eval, new_child = self.minimax(new_board, depth - 1, False)
                # debug start
                #print(f"\nNew {self.player} Board (Grade: {eval}):\n")
                #print(new_board)
                # debug end
                children.append(new_child)
                if eval > max_eval:
                    max_eval = eval
                    best_move = next_moves[new_boards.index(new_board)]
            return best_move, max_eval, TreeNode(children, max_eval, board)
        else:
            min_eval = float('inf')
            next_moves, new_boards = self.calculateMoves(board)
            for new_board in new_boards:
                _, eval, new_child = self.minimax(new_board, depth - 1, True)
                # debug start
                #print(f"\nNew {self.opponent} Board (Grade: {eval}):\n")
                #print(new_board)
                # debug end
                children.append(new_child)
                if eval < min_eval:
                    min_eval = eval
                    best_move = next_moves[new_boards.index(new_board)]
            return best_move, min_eval, TreeNode(children, min_eval, board)
        
    
    def negamax(self, board: Board, depth: int, p: int) -> tuple[int, list[TreeNode]]:
        ''' Negamax algorithm 
        p: 1 for player, -1 for opponent
        '''
        if depth == 0:
            grade = p * self.grade(board)
            return None, grade, TreeNode([], grade, board)
        
        children = []
        max_eval = -float('inf')
        next_moves, new_boards = self.calculateMoves(board)
        for new_board in new_boards:
            _, eval, new_child = self.negamax(new_board, depth - 1, -p)
            # debug start
            #print(f"\nNew {new_board.getTurn()} Board (Grade: {eval}):\n")
            #print(new_board)
            # debug end
            children.append(new_child)
            eval = -eval
            if eval > max_eval:
                max_eval = eval
                best_move = next_moves[new_boards.index(new_board)]
        return best_move, max_eval, TreeNode(children, max_eval, board)
    
    def alphabeta(self, board: Board, depth: int, alpha: int, beta: int, p: int) -> tuple[int, list[TreeNode]]:
        ''' Alpha-Beta Pruning algorithm 
        p: 1 for player, -1 for opponent
        '''
        if depth == 0:
            value = p * self.grade(board)
            return None, value, TreeNode([], value, board)
        
        children = []
        value = alpha
        best_moves: list[Move] = []
        next_moves, new_boards = self.calculateMoves(board)
        for new_board in new_boards:
            _, eval, new_child = self.alphabeta(new_board, depth - 1, -beta, -value, -p)
            children.append(new_child)
            eval = -eval
            if eval > value:
                value = eval
                best_moves = [next_moves[new_boards.index(new_board)]]
                # debug start
                #print(f"\nNew {new_board.getTurn()} Board (Grade: {eval}):\n")
                #print(new_board)
                # debug end
            elif eval == value:
                best_moves.append(next_moves[new_boards.index(new_board)])
            if value >= beta:
                if best_moves:
                    chosen_move = random.choice(best_moves)
                else:
                    chosen_move = GameEnd()
                return chosen_move, value, TreeNode(children, value, board)
        if best_moves:
            chosen_move = random.choice(best_moves)
        else:
            chosen_move = GameEnd()
        return chosen_move, value, TreeNode(children, value, board)

File: ai.c
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "dynamicarray.c"
#include "gamerep.h"
#include "rules.c"

size_t depthNodes[4] = {0, 0, 0, 0};
size_t totalNodes = 0;

typedef struct fendoterSettings{
    unsigned int searchDepth;
    int playingMethod;
}fendoterSettings;

void evaluateMoves(move_t* bestMove, field_t* boardState, fendoterSettings* settings);
void playRandom(field_t* board, move_t* bestMove);
int minimax(move_t* bestMove);
int negamax(field_t* board, int depth, int p, move_t* bestMove, fendoterSettings* settings);
int alphaBeta(field_t* board, int depth, int alpha, int beta, int p, move_t* bestMove);

/* board operations */
// Note: the do not check if the move is legal
void placePawn(char x, char y, char player, field_t* board){
    char pawn, pawnDataOffset, nextTurn;
    field_t* pawnMetaData;
    if (player == 1){
        pawn = PLAYER1PAWN;
        pawnDataOffset = PAWNS1NUM;
        nextTurn = 2;
    }
    else {
        pawn = PLAYER2PAWN;
        pawnDataOffset = PAWNS2NUM;
        nextTurn = 1;
    }
    field_t* field = board + x + 7*y;
    *field = *field | pawn;

    /* update meta data */
    pawnMetaData = board + pawnDataOffset;
    *pawnMetaData += 1;
    board[TURN] = nextTurn;
}

/* Note wallDirection is already the correct mask*/
void placeWall(char x, char y, char wallDirection, field_t* board){
    field_t *field, *field2;
    field = board + x + 7*y;
    *field = *field | wallDirection | HASWALL;
    switch (wallDirection){
        case WALLNORTH:
            field2 = field - 7;
            *field2 = *field2 | WALLSOUTH | HASWALL;
            break;
        case WALLSOUTH:
            field2 = field + 7;
            *field2 = *field2 | WALLNORTH | HASWALL;
            break;
        case WALLEAST:
            field2 = field + 1;
            *field2 = *field2 | WALLWEST | HASWALL;
            break;
        case WALLWEST:
            field2 = field - 1;
            *field2 = *field2 | WALLEAST | HASWALL;
            break;
    }
    
    /* update meta data */
    board[TURN] = (board[TURN] == 1) ? 2 : 1;
}

void movePawn(char x, char y, char u, char v, char player, field_t* board){
    field_t *startField, *endField;
    char pawn = (player == 1) ? PLAYER1PAWN : PLAYER2PAWN;

    startField = board + x + 7*y;
    endField = board + u + 7*v;
    // Remove pawn from original field
    *startField = *startField & ~OCCUPIED;

    // Add pawn to original field
    *endField = *endField | pawn;
}


move_t* makeMove(field_t* boardState, fendoterSettings* settings){
    printf("Making move\n");
    move_t* move = (move_t*) malloc(sizeof(move_t));
    for (int i = 0; i < 54; i++){
        printf("%x ", boardState[i]);
        if (i % 7 == 6){
            printf("\n");
        }
    }
    evaluateMoves(move, boardState, settings);
    return move;
}

void freeBestMove(move_t* bestMove){
    free(bestMove);
}

void evaluateMoves(move_t* bestMove, field_t* boardState, fendoterSettings* settings){
    switch (settings->playingMethod){
        case RANDOM:
            printf("Using playing method: RANDOM\n");
            playRandom(boardState, bestMove);
            printf("Total Nodes: %zu\n", totalNodes);
            printf("Number of nodes at each depth: %zu, %zu, %zu, %zu\n", depthNodes[0], depthNodes[1], depthNodes[2], depthNodes[3]);
            break;
        case MINIMAX:
            printf("Using playing method: MINIMAX\n");
            minimax(bestMove);
            break;
        case NEGAMAX:
            printf("Using playing method: NEGAMAX\n");
            negamax(boardState, settings->searchDepth, 1, bestMove, settings);
            printf("Number of nodes at each depth: %zu, %zu, %zu, %zu\n", depthNodes[0], depthNodes[1], depthNodes[2], depthNodes[3]);
            break;
        case ALPHABETA:
            printf("Using playing method: ALPHABETA\n");
            alphaBeta(boardState, settings->searchDepth, INT_MIN, INT_MAX, 1, bestMove);
            break;
    }
}


void calculateMoves(field_t* board, dynamic_array_move_t* moves, dynamic_array_ucharp* newBoards){
    char turn = (char) *(board + TURN);
    char curPlayer, curOpponent;
    char x, y, u, v;
    field_t *iField, *kField;

    if (turn == 1){
        curPlayer = PLAYER1PAWN;
        curOpponent = PLAYER2PAWN;
    }
    else {
        curPlayer = PLAYER2PAWN;
        curOpponent = PLAYER1PAWN;
    }

    for (int i = 0; i < 49; i++){
        iField = board + i;
        if ( (*iField & ASSIGNED) || (*iField & curOpponent) ){
            continue;
        }
        if (*iField & curPlayer){
            x = i % 7;
            y = i / 7;
            for (int j = 0; j < 4; j++){
                if (checkWallPlace(x, y, DIRECTIONS[j], board)){
                    field_t* newBoard = (field_t*) malloc(sizeof(field_t) * 54);
                    memcpy(newBoard, board, sizeof(field_t) * 54);
                    placeWall(x, y, DIRECTIONS[j], newBoard);
                    if (checkOpenArea(x, y, DIRECTIONS[j], newBoard)){
                        move_t* move = (move_t*) malloc(sizeof(move_t));
                        addItemUCharP(newBoards, newBoard);
                        move->moveType = PLACEWALL;
                        move->direction = DIRECTIONS[j];
                        move->x = x;
                        move->y = y;
                        move->u = -1;
                        move->v = -1;
                        move->player = turn;
                        addItemMove(moves, move);
                        totalNodes++;
                        continue;
                    }
                    else {
                        free(newBoard);
                    }
                }
            }
            for (int k = 0; k < 49; k++){
                kField = board + k;
                if ( (*kField & ASSIGNED) || (*kField & OCCUPIED) ){
                    continue;
                }
                u = k % 7;
                v = k / 7;
                if (checkPawnMove(x, y, u, v, board)){
                    field_t tempBoard[54];
                    memcpy(tempBoard, board, sizeof(field_t) * 54);
                    movePawn(x, y, u, v, turn, tempBoard);
                    for (int l = 0; l < 4; l++){
                        if (checkWallPlace(u, v, DIRECTIONS[l], tempBoard)){
                            field_t* newBoard = (field_t*) malloc(sizeof(field_t) * 54);
                            memcpy(newBoard, tempBoard, sizeof(field_t) * 54);
                            placeWall(u, v, DIRECTIONS[l], newBoard);
                            if (checkOpenArea(u, v, DIRECTIONS[l], newBoard)){
                                move_t* move = (move_t*) malloc(sizeof(move_t));
                                addItemUCharP(newBoards, newBoard);
                                move->moveType = MOVEPAWNANDWALL;
                                move->direction = DIRECTIONS[l];
                                move->x = x;
                                move->y = y;
                                move->u = u;
                                move->v = v;
                                move->player = turn;
                                addItemMove(moves, move);
                                totalNodes++;
                            }
                            else {
                                free(newBoard);
                            }
                        }
                    }
                }
            }
        }
        else { /* should only be called when field is empty !(*iField & OCCUPIED) */
            u = i % 7;
            v = i / 7;
            if (checkPawnPlace(u, v, turn, board)){
                field_t* newBoard = (field_t*) malloc(sizeof(field_t) * 54);
                move_t* move = (move_t*) malloc(sizeof(move_t));
                memcpy(newBoard, board, sizeof(field_t) * 54);
                placePawn(u, v, turn, newBoard);
                addItemUCharP(newBoards, newBoard);
                move->moveType = PLACEPAWN;
                move->x = -1;
                move->y = -1;
                move->u = u;
                move->v = v;
                move->player = turn;
                addItemMove(moves, move);
                totalNodes++;
            }
        }
    }
}


int checkRightField(int i, field_t* field, char player){
    if (i % 7 == 6){ /* Check for right boarder */
        return -1;
    }
    else if (*field & WALLEAST) { /* Check for wall before occupied */
        return 3;
    }
    else if (*(field + 1) & OCCUPIED) {
        if (*(field + 1) & player) {
            return 1;
        }
        else { /* If neighboring field not occupied by own player that it must be opponent's player */
            return 2;
        }
    }
    else {
        return 0;
    }
}


int checkLeftField(int i, field_t* field, char player){
    if (i % 7 == 0){ /* Check for left boarder */
        return -1;
    }
    else if (*field & WALLWEST) { /* Check for wall before occupied */
        return 3;
    }
    else if (*(field - 1) & OCCUPIED) {
        if (*(field - 1) & player) {
            return 1;
        }
        else { /* If neighboring field not occupied by own player that it must be opponent's player */
            return 2;
        }
    }
    else {
        return 0;
    }
}


int checkTopField(int i, field_t* field, char player){
    if (i < 7){ /* Check for top boarder */
        return -1;
    }
    else if (*field & WALLNORTH) { /* Check for wall before occupied */
        return 3;
    }
    else if (*(field - 7) & OCCUPIED) {
        if (*(field - 7) & player) {
            return 1;
        }
        else { /* If neighboring field not occupied by own player that it must be opponent's player */
            return 2;
        }
    }
    else {
        return 0;
    }
}


int checkBottomField(int i, field_t* field, char player){
    if (i > 41){ /* Check for bottom boarder */
        return -1;
    }
    else if (*field & WALLSOUTH) { /* Check for wall before occupied */
        return 3;
    }
    else if (*(field + 7) & OCCUPIED) {
        if (*(field + 7) & player) {
            return 1;
        }
        else { /* If neighboring field not occupied by own player that it must be opponent's player */
            return 2;
        }
    }
    else {
        return 0;
    }
}


int evaluateBoard(field_t* board){
    int grade = 0, areaGrade = 0, freedomGrade = 0;
    int freedomGradeCurPly = 0, freedomGradeOppPly = 0;
    field_t iField;

    char currentPlayer, opponentPlayer;
    char playerAssignedFields, opponentAssignedFields;

    /* Identify current player */
    char currentTurn = (char) *(board + TURN);
    if (currentTurn == 1){
        currentPlayer = PLAYER1PAWN;
        opponentPlayer = PLAYER2PAWN;
        playerAssignedFields = (char) *(board + ASSIGNEDFIELDS1);
        opponentAssignedFields = (char) *(board + ASSIGNEDFIELDS2);
    }
    else {
        currentPlayer = PLAYER2PAWN;
        opponentPlayer = PLAYER1PAWN;
        playerAssignedFields = (char) *(board + ASSIGNEDFIELDS2);
        opponentAssignedFields = (char) *(board + ASSIGNEDFIELDS1);
    }

    /* Check if current player has won */
    if (playerAssignedFields >= 25){
        return INT_MAX;
    }
    if (opponentAssignedFields >= 25){
        return INT_MIN;
    }

    /* Grade occupied area */
    areaGrade = playerAssignedFields - opponentAssignedFields;

    /* Grade freedom of the pawns */
    
    for (int i = 0; i < 49; i++){
        iField = *(board + i);
        if (iField & currentPlayer){
            switch (checkRightField(i, &iField, currentPlayer)) {
                case -1: /* Boarder to right */
                    freedomGradeCurPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeCurPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeCurPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeCurPly -= 6;
                    break;
                case 3: /* Wall */
                    freedomGradeCurPly -= 10;
                    break;
            }
            switch (checkLeftField(i, &iField, currentPlayer)){
                case -1: /* Boarder to left */
                    freedomGradeCurPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeCurPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeCurPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeCurPly -= 6;
                    break;
            }
            switch (checkTopField(i, &iField, currentPlayer)){
                case -1: /* Boarder to top */
                    freedomGradeCurPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeCurPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeCurPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeCurPly -= 6;
                    break;
            }
            switch (checkBottomField(i, &iField, currentPlayer)){
                case -1: /* Boarder to bottom */
                    freedomGradeCurPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeCurPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeCurPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeCurPly -= 6;
                    break;
            }
        }
        if (iField & opponentPlayer){
            switch (checkRightField(i, &iField, currentPlayer)) {
                case -1: /* Boarder to right */
                    freedomGradeOppPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeOppPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeOppPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeOppPly -= 6;
                    break;
                case 3: /* Wall */
                	freedomGradeOppPly -= 10;
                    break;
            }
            switch (checkLeftField(i, &iField, currentPlayer)){
                case -1: /* Boarder to left */
                    freedomGradeOppPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeOppPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeOppPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeOppPly -= 6;
                    break;
            }
            switch (checkTopField(i, &iField, currentPlayer)){
                case -1: /* Boarder to top */
                    freedomGradeOppPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeOppPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeOppPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeOppPly -= 6;
                    break;
            }
            switch (checkBottomField(i, &iField, currentPlayer)){
                case -1: /* Boarder to bottom */
                    freedomGradeOppPly -= 10;
                    break;
                case 0: /* Empty Field */
                    //freedomGradeOppPly += 1;
                    break;
                case 1: /* Occupied by own pawn */
                    freedomGradeOppPly -= 3;
                    break;
                case 2: /* Occupied by enemy pawn */
                    freedomGradeOppPly -= 6;
                    break;
            }
        }
    }

    freedomGrade = freedomGradeCurPly - freedomGradeOppPly;

    grade = areaGrade + freedomGrade; /* add coefficiants */
    return grade;
}

void playRandom(field_t* board, move_t* bestMove) {
    dynamic_array_move_t* moves;
    dynamic_array_ucharp* newBoards;
    arrayInitMove(&moves);
    arrayInitUCharP(&newBoards);

    srand(time(NULL)); // Seed the random number generator
    calculateMoves(board, moves, newBoards);
    printf("Number of moves: %zu\n", moves->size); // todo: size not correct for "standard move" check that calculate moves works as intended
    size_t placePawnMoves = 0;
    size_t placeWallMoves = 0;
    size_t movePawnAndWallMoves = 0;
    size_t undefinedMoves = 0;
    for (size_t i = 0; i < moves->size; i++) {
        move_t* tempMove = getItemMove(moves, i);
        switch (tempMove->moveType) {
            case PLACEPAWN:
                placePawnMoves++;
                break;
            case PLACEWALL:
                placeWallMoves++;
                break;
            case MOVEPAWNANDWALL:
                movePawnAndWallMoves++;
                break;
            default:
                undefinedMoves++;
                break;
            
        }
    }
    printf("Number of place pawn moves: %zu\n", placePawnMoves);
    printf("Number of place wall moves: %zu\n", placeWallMoves);
    printf("Number of move pawn and wall moves: %zu\n", movePawnAndWallMoves);
    printf("Number of undefined moves: %zu\n", undefinedMoves);
    if (newBoards->size > 0) {
        size_t randomIndex = rand() % newBoards->size;
        move_t* randomMove = getItemMove(moves, randomIndex);
        bestMove->moveType = randomMove->moveType;
        bestMove->x = randomMove->x;
        bestMove->y = randomMove->y;
        bestMove->u = randomMove->u;
        bestMove->v = randomMove->v;
        bestMove->direction = randomMove->direction;
        bestMove->player = randomMove->player;
        printf("Random move is:\n");
        printf("Move type: %x\n", bestMove->moveType);
        printf("x: %d\n", bestMove->x);
        printf("y: %d\n", bestMove->y);
        printf("u: %d\n", bestMove->u);
        printf("v: %d\n", bestMove->v);
        printf("direction: %d\n", bestMove->direction);
        printf("player: %d\n", bestMove->player);
    }
    freeArrayMove(moves);
    freeArrayUCharP(newBoards);
}

int negamax(field_t* board, int depth, int p, move_t* bestMove, fendoterSettings* settings){
    int grade, maxEval, eval;
    dynamic_array_move_t* moves;
    dynamic_array_ucharp* newBoards;
    arrayInitMove(&moves);
    arrayInitUCharP(&newBoards);

    if (depth == 0) {
        grade = p * evaluateBoard(board);
        return grade;
    }

    maxEval = INT_MIN;
    calculateMoves(board, moves, newBoards); // each depth produces the factor of around 200 new boards 186-37856-7618193
    depthNodes[settings->searchDepth - depth] += newBoards->size;

    for (size_t i = 0; i < newBoards->size; i++) {
        eval = -negamax(newBoards->array[i], depth - 1, -p, bestMove, settings);
        if (eval > maxEval) {
            maxEval = eval;
            if (depth == settings->searchDepth){ // override bestMove only in the top most layer
                move_t* tempMove = getItemMove(moves, i);
                bestMove->moveType = tempMove->moveType;
                bestMove->x = tempMove->x;
                bestMove->y = tempMove->y;
                bestMove->u = tempMove->u;
                bestMove->v = tempMove->v;
                bestMove->direction = tempMove->direction;
                bestMove->player = tempMove->player;
                printf("Best move with grade %d is:\n", maxEval);
                printf("Move type: %x\n", bestMove->moveType);
                printf("x: %d\n", bestMove->x);
                printf("y: %d\n", bestMove->y);
                printf("u: %d\n", bestMove->u);
                printf("v: %d\n", bestMove->v);
                printf("direction: %d\n", bestMove->direction);
                printf("player: %d\n", bestMove->player);
            }
        }
    }
    freeArrayMove(moves);
    freeArrayUCharP(newBoards);
    return maxEval;
}


int minimax(move_t* bestMove){
    return 1;
}

int alphaBeta(field_t* board, int depth, int alpha, int beta, int p, move_t* bestMove){
    return 1;
}

// Board representation in cpp?
/*
- array of 53 uint8_t 49 for the fields and 4 for num of pawns1/2, num of assigned fields1/2


- array for moves, append at end









*/





// Rules in cpp (as functions -> rules.cpp)